plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java

```diff
@@ -305,6 +328,109 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+            projections.add(new Variable(pinotColumnHandle.getColumnName(), pinotColumnHandle.getDataType()));
+            resultAssignments.add(new Assignment(pinotColumnHandle.getColumnName(), pinotColumnHandle, pinotColumnHandle.getDataType()));
+        }
+        List<String> groupingColumns = getOnlyElement(groupingSets).stream()
+                .map(PinotColumnHandle.class::cast)
+                .map(PinotColumnHandle::getColumnName)
+                .collect(toImmutableList());
+        OptionalLong limitForDynamicTable = OptionalLong.empty();
+        // Ensure that pinot default limit of 10 rows is not used
+        // By setting the limit to maxRowsPerBrokerQuery + 1 the connector will
+        // know when the limit was exceeded and throw an error
+        if (tableHandle.getLimit().isEmpty() && !groupingColumns.isEmpty()) {
+            limitForDynamicTable = OptionalLong.of(maxRowsPerBrokerQuery + 1);
+        }
+        DynamicTable dynamicTable = new DynamicTable(
+                tableHandle.getTableName(),
+                Optional.empty(),
+                ImmutableList.of(),
+                tableHandle.getQuery().flatMap(DynamicTable::getFilter),
+                groupingColumns,
+                aggregationExpressions.build(),
+                ImmutableList.of(),
+                limitForDynamicTable,
+                OptionalLong.empty(),
+                "");
+        tableHandle = new PinotTableHandle(tableHandle.getSchemaName(), tableHandle.getTableName(), tableHandle.getConstraint(), tableHandle.getLimit(), Optional.of(dynamicTable));
+
+        return Optional.of(new AggregationApplicationResult<>(tableHandle, projections.build(), resultAssignments.build(), ImmutableMap.of(), false));
+    }
+
+    private Optional<PinotColumnHandle> applyCountDistinct(ConnectorSession session, AggregateFunction aggregate, Map<String, ColumnHandle> assignments, PinotTableHandle tableHandle, Optional<PinotColumnHandle> rewriteResult)
+    {
+        AggregateFunctionRule.RewriteContext context = new AggregateFunctionRule.RewriteContext()
+        {
+            @Override
+            public Map<String, ColumnHandle> getAssignments()
+            {
+                return assignments;
+            }
+
+            @Override
+            public Function<String, String> getIdentifierQuote()
+            {
+                return identity();
+            }
+
+            @Override
+            public ConnectorSession getSession()
+            {
+                return session;
+            }
+        };
+
+        if (implementCountDistinct.getPattern().matches(aggregate, context) && aggregate.getInputs().size() == 1) {
+            Variable input = (Variable) getOnlyElement(aggregate.getInputs());
+            if (tableHandle.getQuery().isEmpty() ||
+                    !tableHandle.getQuery().get().getGroupingColumns().contains(input.getName())) {
+                return Optional.empty();
```

Comment:
Should this return empty if the grouping column doesn't contain the column on which aggregation is applied? Or should it throw an error?

Reply:
It should return empty not push the aggregation down. Added tests to verify this behavior with corner cases, like when the broker query is present and there is no count distinct. Also added a comment explaining the reason: count(<column name>) is allowed in the query, just should not be pushed down, it's not an error condition.